source/common/network/dns_resolver/dns_factory.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/config/cluster/v3/cluster.pb.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.pb.h"
+#include "envoy/extensions/filters/udp/dns_filter/v3alpha/dns_filter.pb.h"
+#include "envoy/extensions/network/dns_resolver/apple/v3/apple_dns_resolver.pb.h"
+#include "envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.pb.h"
+#include "envoy/network/dns.h"
+
+#include "source/common/config/utility.h"
+#include "source/common/runtime/runtime_features.h"
+
+namespace Envoy {
+namespace Network {
+
+const std::string cares_dns_resolver = "envoy.dns_resolver.cares";
```

Comment:
`constexpr absl::string_view`

Reply:
I tried to change this into absl::string_view, but hitting multiple compiling issues, for example: 
"candidate function not viable: no known conversion from 'const absl::string_view' to 'const std::string'"
These are due to : 
1)  in the below factory lookup, which is using std::string as parameter:
template <class Factory> static Factory& getAndCheckFactoryByName(const std::string& name) 

2)  The "name" in TypedExtensionConfig is a std::string.

3) The return type of name() member function in class UntypedFactory which the DnsResolverFactory() is based on is a std::string.  

Therefore, let's keep it as std::string to keep them consistent.

Reply:
also, std::string can not be constexpr due to:

./source/common/network/dns_resolver/dns_factory.h:20:23: error: constexpr variable cannot have non-literal type 'const std::string' (aka 'const basic_string<char, char_traits<char>, allocator<char>>')
constexpr std::string CaresDnsResolver = "envoy.network.dns_resolver.cares";

Thus let me keep them as const std::string.

Reply:
use absl::string_view and if necessary change call-sites like getAndCheckFactoryByName to also use absl::string_view.

It's fine to convert from absl::string_view to std::string in the implementation of name(). A copy will be made on the return from that function in any case.

It is not OK to have a static std::string -- it's a static-init fiasco.